src/main/java/org/mule/module/json/internal/cleanup/JsonModuleResourceReleaser.java

```diff
@@ -0,0 +1,272 @@
+      LOGGER.warn("Caught exception while stopping the Executor Service references of the MessageBundles class: {}",
+                  ex.getMessage(), ex);
+    } finally {
+      if (bundleField != null) {
+        bundleField.setAccessible(isAccessible);
+      }
+    }
+
+    bundleField = null;
+    isAccessible = false;
+
+    try {
+      bundleField = ProcessingMessage.class.getDeclaredField("BUNDLE");
+      isAccessible = bundleField.isAccessible();
+      bundleField.setAccessible(true);
+      MessageBundle messageBundle = (MessageBundle) bundleField.get(null);
+      cleanMessageBundle(messageBundle);
+    } catch (NoSuchFieldException | IllegalAccessException | InterruptedException ex) {
+      LOGGER.warn("Caught exception while stopping the Executor Service references of the ProcessingMessage class: {}",
+                  ex.getMessage(), ex);
+    } finally {
+      if (bundleField != null) {
+        bundleField.setAccessible(isAccessible);
+      }
+    }
+
+  }
+
+  /**
+   * Stops the Executor Service instances of the MessageBundle class
+   * @param bundle a MessageBundle Instance
+   * @throws NoSuchFieldException the requested field does not exists
+   * @throws InterruptedException the executor shutdown was interrupted due timeout
+   */
+  private void cleanMessageBundle(MessageBundle bundle)
+      throws NoSuchFieldException, IllegalAccessException, InterruptedException {
+
+    Field providersField = null;
+    Field serviceField = null;
+    boolean isProviderFieldAccessible = false;
+    boolean isServiceFieldAccessible = false;
+
+    try {
+      providersField = MessageBundle.class.getDeclaredField("providers");
+      isProviderFieldAccessible = providersField.isAccessible();
+      providersField.setAccessible(true);
+
+      List<MessageSourceProvider> messageSourceProviders = (List<MessageSourceProvider>) providersField.get(bundle);
+      for (MessageSourceProvider provider : messageSourceProviders) {
+        if (provider instanceof LoadingMessageSourceProvider) {
+          try {
+            serviceField = LoadingMessageSourceProvider.class.getDeclaredField("service");
+            isServiceFieldAccessible = serviceField.isAccessible();
+            serviceField.setAccessible(true);
+            ExecutorService service = (ExecutorService) serviceField.get(provider);
+            service.shutdown();
+            service.awaitTermination(10, TimeUnit.SECONDS);
+          } finally {
+            if (serviceField != null) {
+              serviceField.setAccessible(isServiceFieldAccessible);
+            }
+            serviceField = null;
+            isServiceFieldAccessible = false;
+          }
+        }
+      }
+    } finally {
+      if (providersField != null) {
+        providersField.setAccessible(isProviderFieldAccessible);
+      }
+    }
+  }
+
+  /**
+   * Restores de ExecutorServices for the underlying library.
+   */
+  public synchronized void restoreExecutorServices() {
```

Comment:
Just a thought, this method and the releaseExecutors method look quite similar, is there a way that we can refactor this to reuse code?

Reply:
The release executor shutdowns the executors... the restoreExecutors replaces them with instances obtained through the SchedulerService.... perhaps some improvement can be achieved.